# Exam Review

## Roots

### Bracketing Methods
You must know that the root lies between the two brackets. If you evaluated the function at each value, there should be a difference in sign between the values for there to be a root in between them.

#### Bisection

Take an average $x_r$, between the lower bracket value, $x_l$, and the upper bracket value, $x_u$ then evaluate the function with that average value as follows:

Evaluate $f(x_l)$ and $f(x_u)$ to get values. Check for sign differences.

Find the average

$$x_r=\frac{x_l+x_u}{2}$$

Evaluate $f(x_r)$, if its value is zero (or very close) then $x_r$ is the root. Otherwise let it become the new $x_l$ or $x_u$ depending on which one has the opposite sign of the $x_r$. Repeat until the error is below the desired value.

Error is calculated as

$$\varepsilon_a=\left| \frac{x_r^{new}-x_r^{old}}{x_r^{new}} \right | 100\%$$

##### Example
Find the root of the following function:

$$f(x)=x^{10}-1$$ 

between $x=0$ and $x=1.3$

Therefore $x_l=0$ and $x_u=1.3$. So calculate the values first.

$$f(x_l)=f(0)=0^{10}-1=-1$$
and
$$f(x_u)=f(0)=1.3^{10}-1=12.785849...$$

In [60]:
xl = 0
xu = 1.3

In [61]:
def f(x): return x**10-1

In [62]:
f(xl)

-1

In [63]:
f(xu)

12.785849184900005

Signs are different (-1 and +12.785...) so a root lies between. Calculated the average value $x_r$.

$$x_r=\frac{x_l+x_u}{2}=\frac{0+1.3}{2}=0.65$$

In [64]:
xr = (xl+xu)/2
xr

0.65

Evaluate the function at $x_r$

$$f(x_r)=f(0.65)=0.65^{10}-1=-0.9865...$$

In [67]:
f(xr)

-0.9865372566553711

The value is negative so it becomes the new $x_l$ ($x_u$ is positive i.e. the opposite sign). 

Error is evaluated as (use the $x_l=0$ for the first calculation


$$\varepsilon_a=\left| \frac{0.65-0}{0.65} \right | 100\%=100\%$$

In [68]:
err = abs((xr-xl)/xr)*100
err

100.0

Repeat the process until error is sufficiently small.

In [69]:
i = 2
while err > 0.01:
    if f(xr)<0:
        xl = xr
    else:
        xu = xr
    xr = (xl+xu)/2
    err = abs((xr-xl)/xr)*100
    print('Iteration= ',i,'xr=',xr,'err=',err,'f(xr)=',f(xr))
    i += 1

Iteration=  2 xr= 0.9750000000000001 err= 33.333333333333336 f(xr)= -0.22367037914356147
Iteration=  3 xr= 1.1375000000000002 err= 14.28571428571429 f(xr)= 2.626720217225217
Iteration=  4 xr= 1.0562500000000001 err= 7.692307692307695 f(xr)= 0.7284913860640538
Iteration=  5 xr= 1.015625 err= 3.999999999999991 f(xr)= 0.1677068465489142
Iteration=  6 xr= 0.9953125 err= 2.0408163265306074 f(xr)= -0.04589848926847229
Iteration=  7 xr= 1.00546875 err= 1.010101010100997 f(xr)= 0.05605314088389357
Iteration=  8 xr= 1.000390625 err= 0.507614213197974 f(xr)= 0.003913123612528047
Iteration=  9 xr= 0.9978515625000001 err= 0.25445292620865356 f(xr)= -0.0212778502779869
Iteration=  10 xr= 0.9991210937500001 err= 0.12706480304956191 f(xr)= -0.00875438241808546
Iteration=  11 xr= 0.9997558593750001 err= 0.06349206349206127 f(xr)= -0.0024387257864679768
Iteration=  12 xr= 1.0000732421875 err= 0.031735956839092035 f(xr)= 0.0007326633209661093
Iteration=  13 xr= 0.9999145507812501 err= 0.0158704967465503

| Iteration  | $x_l$  | $x_u$  | $x_r$  | $\varepsilon_a(\%)$  |
|---|---|---|---|---|
|  1 | 0  | 1.3  | 0.65  | 100  |
|  2 | 0.65  | 1.3  | 0.97  | 33.33  |
|  3 |  0.975 | 1.3  | 1.1375  | 14.285  |
|  4 |  0.975 | 1.375  | 1.05625  | 7.692  |
|  5 |  0.975 | 1.05625  | 1.015625  | 4.0  |
|  ... |  ... | ...  | ... | ...  |
|  14 |  0.99999... | 1.00000...  | 0.999999...  | 0.0079  |

The final answer is close to zero.

In [70]:
f(xr)

-6.103347989572239e-05

### Open Methods
Only needs one initial guess, however it can get "lost" if not near the root.

#### Newton-Raphson Method

Use the initial guess $x_i$ to find the next value $x_{i+1}$ with the following equation:

$$x_{i+1}=x_i-\frac{f(x_i)}{f'(x_i)}$$

Find the root from the previous function above 

$$f(x)=x^{10}-1$$

Use the initial guess $x_i=1.3$ to avoid a divide by zero situation.

First find the derivitive of the function.

$$f'(x)=10x^9$$

Then the new value can be calculated using

$$x_{i+1}=x_i-\frac{x_i^{10}-1}{10x_i^9}$$

$$x_{i+1}=1.3-\frac{1.3^{10}-1}{10(1.3)^9}=1.17942...$$

In [81]:
xi = 1.3
def fprime(x): return 10*x**9

In [82]:
def NR(xi): return xi-f(xi)/fprime(xi)

In [83]:
xi1 = NR(1.3)
xi1

1.1794299595372328

Repeat process until error is small enough ($x_{i+1}\rightarrow x_i$).

In [84]:
diff = 1.0
while diff>0.0001:
    xi1 = NR(xi)
    diff = abs(xi-xi1)
    print('xi=',xi,'xi1=',xi1,'Difference=',diff)
    xi = xi1
    

xi= 1.3 xi1= 1.1794299595372328 Difference= 0.1205700404627672
xi= 1.1794299595372328 xi1= 1.08413083089608 Difference= 0.09529912864115286
xi= 1.08413083089608 xi1= 1.0240530839332134 Difference= 0.06007774696286661
xi= 1.0240530839332134 xi1= 1.0023894537457745 Difference= 0.021663630187438887
xi= 1.0023894537457745 xi1= 1.0000254692028134 Difference= 0.0023639845429610506
xi= 1.0000254692028134 xi1= 1.0000000029187888 Difference= 2.5466284024666663e-05


## Systems of Linear Equations

### Matrix Methods

A set of equations:

$$a_{11}x_1 + a_{12}x_2 + a_{13}x_3 = b_1$$

$$a_{21}x_1 + a_{32}x_2 + a_{33}x_3 = b_2$$

$$a_{31}x_1 + a_{32}x_2 + a_{33}x_3 = b_3$$

These can be represented by the matrix form:

$$\begin{bmatrix} 
a_{11} & a_{12} & a_{13}\\
a_{21} & a_{22} & a_{23}\\
a_{31} & a_{32} & a_{33}\\
\end{bmatrix}
\begin{bmatrix} 
x_{1}\\
x_{2}\\
x_{3}\\
\end{bmatrix}=\begin{bmatrix} 
b_{1}\\
b_{2}\\
b_{3}\\
\end{bmatrix}
$$

Simplified this is:

$$\bf{A}\cdot x = b$$

Solving as follows

$$\bf{A^{-1}}A\cdot x = A^{-1}b$$

$$\bf x =  A^{-1}b$$




#### Example

The set of equations:

$$1x_1 + 2x_2 + 3x_3 = 4$$

$$5x_1 + 6x_2 + 7x_3 = 8$$

$$2x_1 + 3x_2 + 11x_3 = 1$$

Can be represented as:

$$\begin{bmatrix} 
1 & 2 & 3 &|& 4 \\
5 & 6 & 7 &|& 8 \\
2 & 3 & 11 &|& 1 \\
\end{bmatrix}$$

OR

$$\begin{bmatrix} 
1 & 2 & 3  \\
5 & 6 & 7 \\
2 & 3 & 11 \\
\end{bmatrix}
\begin{bmatrix} 
x_{1}\\
x_{2}\\
x_{3}\\
\end{bmatrix}=\begin{bmatrix} 
4\\
8\\
1\\
\end{bmatrix}
$$


### LU Decomposition

We can decompose the original matrix into a lower and upper triangular matrix and then solve using a forward and backward substitution as follows:

$$\begin{bmatrix} 
a_{11} & a_{12} & a_{13}  \\
a_{21} & a_{22} & a_{23} \\
a_{31} & a_{32} & a_{33} \\
\end{bmatrix} \rightarrow 
\begin{bmatrix} 
1 & 0 & 0  \\
l_{21} & 1 & 0 \\
l_{31} & l_{32} & 1 \\
\end{bmatrix}
\begin{Bmatrix} 
d_{1}\\
d_{2}\\
d_{3}\\
\end{Bmatrix}=
\begin{Bmatrix} 
b_{1}\\
b_{2}\\
b_{3}\\
\end{Bmatrix} \rightarrow
\begin{bmatrix} 
u_{11} & u_{12} & u_{13}  \\
0 & u_{22} & u_{23} \\
0 & 0 & u_{33} \\
\end{bmatrix}
\begin{Bmatrix} 
x_{1}\\
x_{2}\\
x_{3}\\
\end{Bmatrix}=\begin{Bmatrix} 
d_{1}\\
d_{2}\\
d_{3}\\
\end{Bmatrix}=
\begin{Bmatrix} 
x_{1}\\
x_{2}\\
x_{3}\\
\end{Bmatrix}
$$

Decompose A into L and U

$$\bf A = LU$$

Use L and b to find d

$$\bf Ld=b \rightarrow d=L^{-1}b$$

Use d and U to find x

$$\bf Ux=d \rightarrow x=U^{-1}d$$

#### Example

$$\begin{bmatrix} 
1 & 2 & 3 \\
5 & 6 & 7\\
2 & 3 & 11\\
\end{bmatrix}
\begin{bmatrix} 
x_{1}\\
x_{2}\\
x_{3}\\
\end{bmatrix}=\begin{bmatrix} 
4\\
8\\
1\\
\end{bmatrix}
$$

$$x = \begin{bmatrix} 
-2.571428\\
4.142857\\
-0.5714285\\
\end{bmatrix}$$

In [267]:
import numpy as np
A = np.matrix([[1,2,3],[5,6,7],[2,3,11]])
b = np.matrix([[4],[8],[1]])

In [268]:
from scipy.linalg import inv
x = inv(A)*b
print(x)

[[-2.57142857]
 [ 4.14285714]
 [-0.57142857]]


In [269]:
A*x

matrix([[4.],
        [8.],
        [1.]])

## Differentiation

### Central Difference

#### First Order Accuracy 
$$f'(x) \approx \frac{f(x_{i+1})-f(x_{i-1})}{2h}$$

#### Second Order Accuracy
$$f'(x) \approx \frac{-f(x_{i+2})+8f(x_{i+1})-8f(x_{i-1})+f(x_{i-2})}{12h}$$

#### Example
Solve for $f(x)=x^3$ at $x=2$ with $h=0.1$ using the First Order and Second Order Accurate Central Difference Equations.

$$x_i=2$$

$$x_{i+1}=x_i+h=2+0.1=2.1$$

$$x_{i-1}=x_i-h=2-0.1=1.9$$

$$x_{i+2}=x_{i+1}+h=2.1+0.1=2.2$$

$$x_{i-2}=x_{i-1}-h=1.9+-0.1=1.8$$

##### First Order

$$f'(2) \approx \frac{(2.1)^3-(1.9)^3}{2(0.1)}=12.01$$

##### Second Order

$$f'(2) \approx \frac{-(2.2)^3+8(2.1)^3-8(1.9)^3+(1.8)^3}{12(0.1)} = 12.00000000...$$


In [178]:
def f(x): return x**3

In [179]:
xi = 2
h = 0.1
xi1 = xi + h
xi2 = xi1 + h
xim1 = xi - h
xim2 = xim1 - h

In [181]:
def CDf1(xi1,xim1,h): return (f(xi1)-f(xim1))/(2*h)
def CDF2(xi2,xi1,xim1,xim2,h): return (-f(xi2)+8*f(xi1)-8*f(xim1)+f(xim2))/(12*h)

In [182]:
CDf1(xi1,xim1,h)

12.010000000000009

In [183]:
CDF2(xi2,xi1,xim1,xim2,h)

12.000000000000005

## Integration

### Trapezoidal Rule

$$\int_a^b{f(x)}\approx (b-a)\frac{f(a)+f(b)}{2}$$

### Simpson's 1/3 Rule

$$\int_a^b{f(x)}\approx (b-a)\frac{f(x_0)+4f(x_1)+f(x_2)}{6}$$

### Simpson's 3/8 Rule

$$\int_a^b{f(x)}\approx (b-a)\frac{f(x_0)+3f(x_1)+3f(x_2)+f(x_3)}{8}$$

#### Example

Integrate from $x=0\rightarrow2$ the function $f(x)=3x^2$

or $$\int_0^23x^2$$

##### Trapezoidal Rule

$$\int_0^2{3x^2}\approx (2-0)\frac{3(0)^2+3(2)^2}{2}=12.$$

In [199]:
def f(x): return 3*x**2
def trap(a,b): return (b-a)*(f(a)+f(b))/2

In [200]:
a = 0
b = 2
trap(a,b)

12.0

##### Simpson's 1/3 Rule

$$\int_0^2{3x^2}\approx (2-0)\frac{3(0)^2+4(3(1)^2)+3(2)^2}{6}=8$$

In [208]:
c = (a+b)/2
def simp13(a,b,c): return (b-a)*(f(a)+4*f(c)+f(b))/6

In [209]:
simp13(a,b,c)

8.0

##### Simpson's 3/8 Rule

$$\int_0^2{3x^2}\approx (2-0)\frac{3(0)^3+3(3(0.6666)^3+3(3(1.3333)^3)+3(2)^3}{8}=8$$

In [215]:
c = (a+b)/3
d = 2*c
def simp38(a,b,c,d): return (b-a)*(f(a)+3*f(c)+3*f(d)+f(b))/8

In [216]:
simp38(a,b,c,d)

8.0